python/smqtk/web/nn_descriptor_service/__init__.py

```diff
@@ -0,0 +1,231 @@
+    """
+    Simple server that takes in a specification of the following form:
+
+        /nn/<path:uri>[?...]
+
+    Computes the nearest neighbor index for the given data and returns a list
+    of nearest neighbors in the following format
+
+    Standard return JSON:
+    {
+        "success": <bool>,
+        "neighbors": [ <float>, ... ]
+        "message": <string>,
+        "reference_uri": <uri>
+    }
+
+    # Additional Configuration
+
+    ## Environment variable hook
+    We will look for an environment variable `DescriptorService_CONFIG` for a
+    string file path to an additional JSON configuration file to consider.
+
+    """
+
+    @classmethod
+    def is_usable(cls):
+        return True
+
+    @classmethod
+    def get_default_config(cls):
+        """
+        Generate and return a default configuration dictionary for this class.
+        This will be primarily used for generating what the configuration
+        dictionary would look like for this class without instantiating it.
+
+        :return: Default configuration dictionary for the class.
+        :rtype: dict
+
+        """
+        c = super(NearestNeighborServiceServer, cls).get_default_config()
+        merge_configs(c, {
+            "descriptor_factory": DescriptorElementFactory.get_default_config(),
+            "descriptor_generators": {
+                "example": plugin.make_config(get_descriptor_generator_impls)
+            },
+            "nn_index": plugin.make_config(get_nn_index_impls),
+        })
+        return c
+
+    def __init__(self, json_config):
+        """
+        Initialize application based of supplied JSON configuration
+
+        :param json_config: JSON configuration dictionary
+        :type json_config: dict
+
+        """
+        super(NearestNeighborServiceServer, self).__init__(json_config)
+
+        # Descriptor factory setup
+        self.log.info("Initializing DescriptorElementFactory")
+        self.descr_elem_factory = DescriptorElementFactory.from_config(
+            self.json_config['descriptor_factory']
+        )
+
+        # Descriptor generator configuration labels
+        #: :type: dict[str, dict]
+        self.generator_label_configs = self.json_config['descriptor_generator']
+
+        # Cache of DescriptorGenerator instances so we don't have to continuously
+        # initialize them as we get requests.
+        self.descriptor_cache = {}
+        self.descriptor_cache_lock = multiprocessing.RLock()
+
+        self.nn_index = plugin.from_plugin_config(
+            json_config['nn_index'],
+            get_nn_index_impls
+        )
+
+        self.descriptor_generator_inst = plugin.from_plugin_config(
+                                            self.generator_label_configs,
+                                            get_descriptor_generator_impls)
+
+
+        @self.route("/nn/<path:uri>")
+        def compute_nearest_neighbors(uri):
+            """
+            # Data modes for upload/use
```

Comment:
We're going to want to switch this over to reStructured text format (see the descriptor_service method comments) since the addition of sphinx documentation generation.